cranelift/isle/isle/src/trie.rs

```diff
@@ -196,28 +219,86 @@ impl TrieNode {
         if self.is_empty() {
-            *self = TrieNode::Decision { edges: vec![] };
+            *self = TrieNode::Decision {
+                edges: vec![],
+                last_prio: None,
+                cur_prio: None,
+            };
         }
 
         // We must be a decision node.
-        let edges = match self {
-            &mut TrieNode::Decision { ref mut edges } => edges,
+        let (edges, last_prio, cur_prio) = match self {
+            &mut TrieNode::Decision {
+                ref mut edges,
+                ref mut last_prio,
+                ref mut cur_prio,
+            } => (edges, last_prio, cur_prio),
             _ => panic!("insert on leaf node!"),
         };
 
+        // If we are inserting at a lower prio than in `cur_prio`,
+        // `cur_prio` moves to `last_prio` (and controls our minimum
+        // insertion point) and we initialize `cur_prio` with our
+        // current priority and new max insertion index (of 0).
+        if cur_prio.is_none() || prio < cur_prio.unwrap().prio {
+            *last_prio = *cur_prio;
+            *cur_prio = Some(PrioFrontier {
+                prio,
+                // This is `None` initially but will be updated below
+                // to at least the frontier, as we always reuse or
+                // insert an edge and that edge is always >=
+                // `last_prio.edge_idx`, if not `None`. It also needs
+                // to be `None` to distinguish the no-edges case from
+                // the one-edge-at-index-0 case: see comment on
+                // frontier increment below.
+                edge_idx: None,
+            });
+        }
+        let cur_prio = cur_prio.as_mut().unwrap();
+
+        // Determine the minimum edge index under which we can insert
+        // while respecting priorities.
+        let start = last_prio.and_then(|frontier| frontier.edge_idx);
+
         // Now find or insert the appropriate edge.
         let edge = edges
             .iter()
-            .position(|edge| edge.symbol == op && edge.prio == prio)
+            .skip(start.unwrap_or(0))
+            .position(|edge| edge.symbol == op)
+            .map(|pos| pos + start.unwrap_or(0))
             .unwrap_or_else(|| {
-                edges.push(TrieEdge {
-                    prio,
-                    symbol: op,
-                    node: TrieNode::Empty,
-                });
-                edges.len() - 1
+                // Insert in a position among our allowed range
+                // (strictly after `start` now) that would be sorted
+                // according to the `op` symbol.
+                let first_after_prev_prio = start.map(|x| x + 1).unwrap_or(0);
+                assert!(first_after_prev_prio <= edges.len());
+                let insert_pos = edges[first_after_prev_prio..]
+                    .binary_search_by(|edge| edge.symbol.cmp(&op))
+                    .unwrap_err()
+                    + first_after_prev_prio;
```

Comment:
There's a problem here with removing the edge at `start` from consideration when inserting a new node: if none of the rules at previous priorities had an op inserted that would sort less than the op we're inserting, the node at `start` may shadow it. Here's the `extractor_ordering_bug.isle` test with the additional rule `(rule 3 (test (identity 3)) 3)` that shows the bug:

```
                                                                                     
(type u32 (primitive u32))                                                           
                                                                                     
(decl identity (u32) u32)                                                            
(extern extractor infallible identity identity)                                      
                                                                                     
(decl is_zero (u32) u32)                                                             
(extern extractor is_zero is_zero)                                                   
                                                                                     
(decl test (u32) u32)                                                                
                                                                                     
(rule 3 (test (identity 3)) 3)                                                       
                                                                                     
;; This exposes a bug where infallible extractors were running before fallible       
;; ones, as the derived ordering for the `Extract` type was ordering them ahead      
;; of the fallible ones. The result is that the fallible `is_zero` extractor         
;; never runs, as the `identity` extractor will always succeed before it's           
;; called.                                                                           
(rule (test (identity x)) x)                                                         
                                                                                     
(rule (test (is_zero x)) 2)           
```

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Some(0)`,
 right: `Some(2)`', /tmp/.tmpSbDsCf/extractor_ordering_bug_main.rs:21:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_run_run_extractor_ordering_bug ... FAILED
```

Reply:
So, the issue with that is that removing the "last frontier plus one" logic for the starting-point also creates a bug: it allows a lower-priority rule to insert an edge prior to an edge that leads to higher-priority leaves.

The fundamental issue is that the sorting of match ops within a priority is at odds with the sharing of the match edges between priorities. The only way for sharing to occur is for the match op in a given prio range to be last, and that same match op to be first in the next prio range.

So there are two possible design regimes:

- Sort order within a prio is absolutely needed for correctness. This is where we were before: the "more specific rule wins" idea. The test case above relies on this: a fallible extractor is more specific than an infallible extractor, and so it must come first.
- Sort order within a prio is arbitrary, and priorities break ties.

My understanding of this whole situation is that by deciding to disallow overlapping rules not disambiguated by priorities, we have eliminated the need for cases like this: we would not allow the ISLE rules above to exist at the same priority, because they overlap.

It was in this context that we realized the need for the "sharing across priority levels" scheme, and this is how this PR arose. In other words, this PR relies on the fact that we are allowed to reorder rules arbitrarily within a priority, and does so in order to maximize sharing.

However, if we *also* still want to have stricter ordering within a priority, and hold this ordering as semantically binding (as the test case above does), then we must abandon this PR, and this optimization scheme generally. This is because we cannot both strictly sort ops within a priority, and share between priorities, unless the last op of a given priority (in sort order) happens to be the same as the first op of the next priority (in sort order).

So: do we ban overlaps, and then use the resulting freedom to merge edges, getting our efficiency back after adding prios to disambiguate overlaps? Or do we go back to the system that we had before, and strictly respect op ordering?